### PR TITLE
[RLlib] Skip updates if some learners don't have batches, prevent deadlocks

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
## Description

With learners, we can arrive in situations where a) we have multiple learners and b) not enough data is produced to feed all learners at a given DDP step. Today, this creates an empty batch with an unbound loss_per_module.

The naive solution is to skip empty batches but that creates deadlocks in torch DDP as workers wait for each other's steps.
I looked into different solutions of how to solve this, such as having blueprint of valid batches to step with if an empty batch arrives but this solution is messy in practice as each learner needs to have the correct bluebrint at all times (even before the first valid training batch arrives).

This PR proposes as solution where, prior to a DDP step, all learners agree that they have data available and how often they will step on any available minibatches.

## Known limitation: differing module sets across Learners

The agreement added here is per Learner, not per module. Each submodule of a `MultiRLModule` is wrapped in its own `TorchDDPRLModule` (`TorchLearner._make_modules_ddp_if_necessary`), so every module has its own all-reduce, and `MultiRLModule._forward_train` only forwards the modules that are present in the batch it is given.

If two Learners receive batches with different sets of module IDs — for example in multi-agent setups where an agent does not appear in every shard of the sampled episodes — the Learner that is missing a module never runs that module's backward pass, while its peers wait on that module's all-reduce, and the group deadlocks. `_should_skip_update` cannot express this case: it is a single boolean per Learner.

This is pre-existing behaviour that this PR neither introduces nor makes worse; it was confirmed on two CPU-DDP Learners fed `{p0, p1}` and `{p0}` respectively.

A follow-up can close it: `self.module.keys()` is identical on every Learner, so the Learners can agree on a per-module bitmask (AND'ed across the group) and train the intersection, filtering the batch to it. The all-or-nothing skip in this PR then becomes the special case of an empty intersection.

### CODE REVIEW

File: rllib/core/learner/torch/torch_differentiable_learner.py
Comment (ArturNiederfahrenhorst) [inline rllib/core/learner/torch/torch_differentiable_learner.py:1]: Changes in this file are not related to the PR

File: rllib/core/learner/torch/torch_meta_learner.py [hunk 1]
```diff
@@ -34,7 +34,7 @@
     TensorType,
 )
 
-logger = logging.getLogger("__name__")
+logger = logging.getLogger(__name__)
```
Comment (ArturNiederfahrenhorst) [inline rllib/core/learner/torch/torch_meta_learner.py:37]: Unrelated bug fix.

File: rllib/core/learner/differentiable_learner.py [hunk 2]
```diff
@@ -407,7 +415,7 @@ def _create_iterator_if_necessary(
         minibatch_size: Optional[int] = None,
         shuffle_batch_per_epoch: bool = False,
         **kwargs,
-    ) -> Iterable:
+    ) -> Optional[Iterable]:
```
Comment (ArturNiederfahrenhorst) [inline rllib/core/learner/differentiable_learner.py:418]: Before, we'd always return an empty dict.
That's not idiomatic in python.
If we are not able create some iterable, let's log that issue >here< and return None.

File: rllib/core/learner/learner_group.py [hunk 3]
```diff
@@ -379,14 +379,12 @@ def update(
...
                     return_state=(return_state and i == 0),
-                    **kw,
                     **kwargs,
                 )
-                for i, (td_shard, kw) in enumerate(
+                for i, td_shard in enumerate(
                     training_data.shard(
                         num_shards=len(self),
                         len_lookback_buffer=self.config.episode_lookback_horizon,
-                        **kwargs,
```
Comment (ArturNiederfahrenhorst) [inline rllib/core/learner/learner_group.py:389]: We don't need that anymore as now the number of minibatch steps (which was the only kwarg we used here) is computed in the learners.

**Comment by ArturNiederfahrenhorst** [comment]
Did 9 runs with 2 GPUs each, 3 on master, 3 with the learner sync on this PR and 3 on this PR, but `never_skip_updates` enabled (so no learner sync). Used CartPole with a small network to minimize update time and maximize the possible impact of the sync. Could not measure the impact so I think we are good from a performance standpoint.

File: rllib/core/learner/training_data.py [hunk 4]
```diff
@@ -49,43 +47,32 @@ def shard(
         self,
         num_shards: int,
         len_lookback_buffer: Optional[int] = None,
-        **kwargs,
```
Comment (ArturNiederfahrenhorst) [inline rllib/core/learner/training_data.py:52]: We don't need that anymore as now the number of minibatch steps (which was the only kwarg we used here) is computed in the learners.

File: rllib/utils/minibatch_utils.py
Comment (ArturNiederfahrenhorst) [inline rllib/utils/minibatch_utils.py:1]: The main point of the changes in this file is that we want to know how many minibatches we produce up front. This enables us to agree on a shared plan among the learners.

**Review by pseudo-rnd-thoughts — COMMENTED** [review]
I'll finish my review tomorrow but I thought that Simon fixed the known limitation where in multi-agent settings if all learners don't get all module-ids in a batch then a deadlock can happen. 

Also, could you send me the code to replicate this issue, I want to understand if my Ray Train hang detector can help for this

File: rllib/core/learner/learner.py [hunk 5]
```diff
@@ -1208,11 +1256,92 @@ def _create_iterator_if_necessary(
...
+                # proposes its own count and the group settles on one. (Without
+                # `minibatch_size` the count is `num_epochs` or 1 on every Learner.)
+                if (
+                    not num_total_minibatches
+                    and self.config.num_learners > 1
+                    and minibatch_size
+                ):
+                    num_total_minibatches = MiniBatchCyclicIterator.num_minibatches(
+                        batch, minibatch_size=minibatch_size, num_epochs=num_epochs
+                    )
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply rllib/core/learner/learner.py:1292]: Done

File: rllib/utils/minibatch_utils.py [hunk 6]
```diff
@@ -72,39 +72,83 @@ def __init__(
...
+        if len(module_batch) == 0:
+            # `__iter__` refuses to cycle a batch with nothing in it; raise the same
+            # way here, since the count is computed before the loop starts. A batch
+            # with NO timesteps at all is handled earlier (`Learner._should_skip_update`
+            # skips the update), so reaching this means only some of the modules are
+            # empty, which the iterator cannot make equal-sized minibatches from.
+            raise ValueError(
+                "One of the module batches is empty! Minibatches need "
+                "`minibatch_size` samples from every module_id."
             )
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply rllib/utils/minibatch_utils.py:127]: Done

File: rllib/core/learner/learner.py [hunk 7]
```diff
@@ -1234,6 +1363,61 @@ def _create_iterator_if_necessary(
...
+        Not called if `config.never_skip_update` is set; an empty batch is an error
+        then.
+
+        Args:
+            batch: The train batch of this `update()` call.
+
+        Returns:
+            True to skip this update.
+        """
+        return (
```
Comment (pseudo-rnd-thoughts) [inline rllib/core/learner/learner.py:1392]: Skipping when **any** module has 0 rows means a single rare module now blocks training for every module in the group.

`ShardBatchIterator` splits each module with `ceil(n / num_shards)`, so the last shard is empty whenever a module has few rows. With 4 learners that happens for n ∈ {1, 2, 3, 5, 6, 9}. Repro on 2 learners: `update(batch={p0: 256 rows, p1: 1 row})` → shards `{p0: 128, p1: 1}` / `{p0: 128, p1: 0}` → rank 1 votes to skip → **neither p0 nor p1 trains**, and all 256 p0 rows are thrown away. With `p1: 4` rows both modules train. It also applies with `num_learners=1`, where there's no desync to prevent.

Suggestions (either would help):
- Drop zero-row modules from the batch before counting, and skip only when nothing is left. The iterator then never sees an empty module.
- Balance the split in `ShardBatchIterator` (`np.array_split`-style), so no shard is empty when n ≥ num_shards.

(Dropping a module on one rank only is still the known per-module DDP limitation, so the real fix is a per-module agreement. Still, this rule makes the problem worse than the old "no data at all" rule did.)

File: rllib/core/learner/learner.py [hunk 8]
```diff
@@ -1208,11 +1256,92 @@ def _create_iterator_if_necessary(
...
+                            LEARNER_UPDATE_SKIPPED_EMPTY_BATCH_LIFETIME
+                            if wants_to_skip
+                            else LEARNER_UPDATE_SKIPPED_FOR_PEER_LIFETIME,
+                        ),
+                        1,
+                        reduce="lifetime_sum",
+                    )
+                    self.metrics.log_value(
+                        (ALL_MODULES, LEARNER_ENV_STEPS_DROPPED_ON_SKIP_LIFETIME),
+                        batch.env_steps(),
```
Comment (pseudo-rnd-thoughts) [inline rllib/core/learner/learner.py:1340]: `batch.env_steps()` under-reports here for batches that came through `ShardBatchIterator`: it builds each shard with `MultiAgentBatch(batch_to_send, int(batch_size))`, where `batch_size` is the shard size of whichever module was iterated **last**.

In the `{p0: 256, p1: 1}` repro above, `learner_env_steps_dropped_on_skip_lifetime` reads **1** on each rank, but 128 p0 rows were dropped on each. Since this metric is meant to make skips visible, it would be more reliable to log `sum(len(b) for b in batch.policy_batches.values())` (agent steps), or both.

**Review by pseudo-rnd-thoughts — COMMENTED** [review]
Ok, checking out the PR, claude spotted the following things

**`update(episodes=[])` still raises before the skip check.** `AddOneTsToEpisodesAndTruncate` (`rllib/connectors/learner/add_one_ts_to_episodes_and_truncate.py:120`) does `episodes[0]`, which raises `IndexError` in the PPO/IMPALA/APPO/BC/MARWIL/IQL learner pipelines before `_should_skip_update` runs. The PR's empty-episodes test passes only because the testing connector doesn't include that piece. So the IMPALA case where a package has fewer episode refs than learners (`ShardObjectRefIterator` gives one learner `[]`) is not covered yet. One fix: return an empty `MultiAgentBatch` from `_make_batch_if_necessary` when `episodes == []`, without calling the connector.

Overall: the sync point and the count-only `MiniBatchCyclicIterator` look right to me. 

My understanding is that if you have missing data on one of the learners, then we skip that training step for all learners. While this is a weird case edge and I think the right decision, we need to document this for users. 

Claude is suggesting narrowing it to a per-module payload (rows per module on each rank), with learners taking part with zero gradients for modules they don't have. 

Also, could `_should_skip_update` and `never_skip_update` be marked experimental, so they don't become public API we later have to deprecate?

File: rllib/core/learner/learner.py [hunk 9]
```diff
@@ -1208,11 +1256,92 @@ def _create_iterator_if_necessary(
...
+                # With several Learners, shards may hold different amounts of data, and
+                # the number of minibatches `MiniBatchCyclicIterator` derives from a
+                # shard differs per Learner. Unless the caller fixes `num_total_minibatches`,
+                # each Learner proposes its own count and the group settles on one. (Without
+                # `minibatch_size` the count is `num_epochs` or 1 on every Learner.)
+                if (
+                    not wants_to_skip
+                    and not num_total_minibatches
+                    and self.config.num_learners > 1
+                    and minibatch_size
```
Comment (pseudo-rnd-thoughts) [inline rllib/core/learner/learner.py]: `num_epochs > 1` **without** `minibatch_size` is never synced, and it can hang the group. The comment above ("Without `minibatch_size` the count is `num_epochs` or 1 on every Learner") holds for single-agent but not multi-agent.

In that case every rank proposes 0. After the sync, `minibatch_size = batch.count` (env steps), and `MiniBatchCyclicIterator` works out the count locally as `max_m ceil(num_epochs * n_m / count)`. `n_m` is module rows, so the counts depend on each shard's data.

Repro, 2 gloo learners: `update(batch={p0: 257, p1: 64}, num_epochs=2)` → shards `{p0: 129, p1: 32}` / `{p0: 128, p1: 32}`, `count = 32` on both → 9 vs 8 minibatches → **hang**. The same applies to episodes batches whenever agent steps ≠ env steps (shared policies, agents that terminate early).

Fix: also propose a count when `num_epochs > 1`, using the `minibatch_size` fallback the iterator will actually use (`batch.count`), i.e. move that fallback above this block.

File: rllib/core/learner/learner.py [hunk 10]
```diff
@@ -1105,7 +1127,20 @@ def update(
             **kwargs,
         )
 
+        # `None` means: skip this update. `_create_iterator_if_necessary` has already
+        # warned and counted it; only the per-update bookkeeping remains.
+        if batch_iter is None:
+            self.after_gradient_based_update(timesteps=timesteps or {})
```
Comment (pseudo-rnd-thoughts) [inline rllib/core/learner/learner.py:1137]: Minor: calling `after_gradient_based_update` on the skip path makes the algorithm-specific hooks run with no update behind them. With windowed metrics, their keys still exist after `reduce()` but read NaN. In PPO, `after_gradient_based_update` peeks the KL metric, finds NaN, and `_update_module_kl_coeff` logs the "KL divergence ... is non-finite" warning on every skipped update, which sends users looking for a model problem that isn't there. It might be safer to skip this hook, or to give it a flag so subclasses can tell the update was skipped.
